### ISSUE TITLE
security: JSP output-encoding hardening bundle (7 fixes)

### ISSUE THREAD

**Issue by yingbull** [carlos-emr/carlos#1635]
## Summary

Seven JSP output-encoding fixes bundled into a single PR scope. All items follow the same pattern: wrap a previously-unencoded emission with the already-imported OWASP `Encode.*` helper (or `Pattern.quote` for the regex case). No API changes, no dependency additions, no database migration, no integration impact. Zero runtime-behavior change for well-formed input.

**Total scope**: ~23 lines changed across 7 files. Closes 6 scanner-flagged REVIEW alerts plus 4 adversarial-discovery XSS vectors that share the same encoding-gap pattern. Estimated ~30-40 min of work.

**Companion to issue #1634** — the same low-risk, backwards-compatible shape.

---

## Fix 1 — Encode `docId` attribute emissions in `MultiPageDocDisplay.jsp`

**File**: `src/main/webapp/documentManager/MultiPageDocDisplay.jsp`
**Lines**: 795, 806, 816, 832, 835

### Problem
`docId` flows from request parameter `segmentID` via `curdoc.getDocId()` and is emitted raw into HTML attribute values:
```jsp
<input type="hidden" name="docId" value="<%=docId%>"/>
<a href="…&docId=<%=docId%>">
```
Classic attribute-context XSS sink. CodeQL flagged this cluster as REVIEW because it reported line offsets that landed on adjacent static lines; the underlying taint flow is real.

### Fix
Wrap each emission with the OWASP attribute encoder:
```jsp
value="<%=Encode.forHtmlAttribute(String.valueOf(docId))%>"
```
If `docId` is typed as a numeric (`int`/`Integer`), `String.valueOf` makes the conversion explicit and avoids NPE on nulls; `Encode.forHtmlAttribute` short-circuits purely-numeric input (no-op) so there's no measurable overhead.

### Backwards compatibility
- Numeric `docId` values (the normal case) pass through `Encode.forHtmlAttribute` unchanged — the output is byte-identical.
- Any non-numeric path (if `segmentID` is ever non-int) is already going to break the SQL query; encoding here is purely defensive.
- No import change: `import org.owasp.encoder.Encode` is already present in nearby JSPs; add to this file if absent.

### Acceptance criteria
- [ ] Add `<%@ page import="org.owasp.encoder.Encode" %>` if missing
- [ ] Wrap line 795, 806, 816, 832, 835 emissions with `Encode.forHtmlAttribute(String.valueOf(docId))`
- [ ] Manual smoke: open a multi-page document, navigate pages — all links/forms work
- [ ] Manual XSS test: attempt `segmentID="><script>alert(1)</script>` → no script execution

### Related alerts
Closes 4 `java/XSS` REVIEW alerts on this file.

---

## Fix 2 — Quote regex metacharacters in `RxWriteScript2Action`

**File**: `src/main/java/io/github/carlos_emr/carlos/prescript/pageUtil/RxWriteScript2Action.java`
**Lines**: 610, 687

### Problem
```java
strength.replaceAll(unit, "");
```
`unit` is read from the DrugRef remote drug reference DB — an unsanitized external data source. `replaceAll` interprets its first argument as a regular expression. A malformed `unit` value (e.g., one containing `.*` or catastrophic backtracking patterns like `(a+)+$`) causes ReDoS or unexpected replacement behavior.

### Fix
Choose one of two equivalent fixes (prefer Option A for minimal diff):

**Option A** — quote regex metacharacters:
```java
strength.replaceAll(Pattern.quote(unit), "");
```

**Option B** — switch to literal replace (no regex involved):
```java
strength.replace(unit, "");
```
Java's `String.replace(CharSequence, CharSequence)` is literal and global — semantically equivalent to `replaceAll` on a quoted pattern, but with clearer intent.

### Backwards compatibility
- For normal `unit` values ("mg", "mL", "IU"), both options produce identical output to the current code.
- Only pathological `unit` values (regex metachars) behave differently — and in those cases the current code is the bug.
- No import change needed for Option B. Option A requires `import java.util.regex.Pattern` (likely already imported via `Pattern.compile` usage in the same class).

### Acceptance criteria
- [ ] Apply Option A or B on lines 610 and 687 (same fix for both)
- [ ] Existing unit tests for `RxWriteScript2Action` pass
- [ ] Manual smoke: prescribe a medication, verify the strength/unit rendering is unchanged

### Related alerts
Closes 2 `java/RegexInjection` REVIEW alerts.

---

## Fix 3 — Encode raw emissions in `unlockAjax.jsp`

**File**: `src/main/webapp/casemgmt/unlockAjax.jsp`
**Lines**: 56, 74, 94, 96, 108, 121

### Problem
The note-unlock AJAX fragment emits PHI and clinical note fields raw into an HTML response that the client loads via `$(id).update(html)` (Prototype.js → `innerHTML` sink):
- Lines 56, 74: `<c:out escapeXml="false" value="${fmtTxt}"/>` where `fmtTxt = note.getNote().replaceAll("\n","<br>")`
- Lines 94, 96: `out.print("<li>" + provider.getFormattedName() + "; ")`
- Line 108: `"&quot;" + note.getEncounter_type() + "&quot;"`
- Line 121: `iss.getIssue().getDescription()` emitted raw inside `<li>`

Clinician-editable note bodies containing `<script>` or `<img onerror>` execute in any subsequent user's session when they open/unlock the note. Real stored XSS discovered via adversarial audit of the assumption underlying the `carlos-ajax.js` FP dismissal.

The sibling fragment `casemgmt/ChartNotesAjax.jsp` already uses `Encode.forHtml()` consistently — this file was missed during the earlier encoding sweep.

### Fix
Mirror the `ChartNotesAjax.jsp` pattern:

```jsp
<%-- Line 56/74: encode BEFORE newline-to-<br> substitution --%>
<%
pageContext.setAttribute("fmtTxt",
    Encode.forHtml(note.getNote()).replaceAll("\n", "<br>"));
%>
<c:out escapeXml="false" value="${fmtTxt}"/>

<%-- Line 94/96 --%>
out.print("<li>" + Encode.forHtml(provider.getFormattedName()) + "; ");

<%-- Line 108 --%>
"&quot;" + Encode.forHtml(note.getEncounter_type()) + "&quot;"

<%-- Line 121 --%>
<li><%=Encode.forHtml(iss.getIssue().getDescription())%></li>
```

### Backwards compatibility
- Benign alphanumeric clinical text renders identically (no visible change to users).
- HTML metacharacters in legitimate note content (`<`, `>`, `&`) are correctly escaped — this is the intended behavior; unencoded HTML in clinical notes is never valid.
- The `replaceAll("\n","<br>")` stays in place; encoding happens first so that `<br>` tags are introduced after encoding and survive to the browser.

### Acceptance criteria
- [ ] Add `<%@ page import="org.owasp.encoder.Encode" %>` if not already present
- [ ] Apply encoding to all 6 sinks above
- [ ] Manual smoke: lock a note, unlock it from another session, verify display renders correctly
- [ ] Manual XSS test: create a note with body `<script>alert(1)</script>`, unlock it → no alert fires; literal text displays

### Related alerts
This sink was **not** in the original scanner output. Discovered via adversarial audit of the JS-layer FP classifications. No scanner alert to close, but this is a real stored-XSS PHI vector that should be addressed.

---

## Fix 4 — Encode OAuth client metadata in `clients.jsp`

**File**: `src/main/webapp/provider/clients.jsp`
**Lines**: 156, 160

### Problem
```jsp
<%= client.getName() %>
<%=srt.getVerifier()%>
```
`client.getName()` is the OAuth `ServiceClient` display name (admin-editable via OAuth-provider registration). `srt.getVerifier()` is the OAuth 1.0a verifier code (server-generated but echoed back in the admin UI). Raw emission into HTML body context.

### Fix
```jsp
<%= Encode.forHtml(client.getName()) %>
<%= Encode.forHtml(srt.getVerifier()) %>
```

### Backwards compatibility
- Client names are typically plain ASCII (vendor name or integration label) — encoded output identical to current.
- The verifier is a server-generated alphanumeric token — encoded output identical.
- No functional change.

### Acceptance criteria
- [ ] Wrap line 156 and 160 emissions with `Encode.forHtml()`
- [ ] Add OWASP encoder import if not present
- [ ] Manual test: admin OAuth client list still renders

### Related alerts
Adversarial discovery. Not in scanner output but same pattern as scanner-flagged sites.

---

## Fix 5 — Encode eform metadata in `fieldnoteselect.jsp`

**File**: `src/main/webapp/eform/fieldNoteReport/fieldnoteselect.jsp`
**Lines**: 108, 111, 113, 141, 144, 146

### Problem
```jsp
<td><%=fieldNoteEform.getFormName()%></td>
<td><%=fieldNoteEform.getSubject()%></td>
<td><%=fieldNoteEform.getFileName()%></td>
…
<a href="fieldNoteReport.do?formName=<%=fieldNoteEform.getFormName()%>…">
```
Admin-editable eform metadata (formName/subject/fileName) emitted raw into table cells and into an `href` attribute. Administrators can craft an eform name with HTML to deliver stored XSS to other admins viewing the list.

### Fix
Use `Encode.forHtml()` for body-context emissions and `Encode.forHtmlAttribute()` for the `href` context:

```jsp
<%-- Body context (lines 108, 111, 113, 141, 144) --%>
<td><%= Encode.forHtml(fieldNoteEform.getFormName()) %></td>

<%-- Attribute context (line 146) --%>
<a href="fieldNoteReport.do?formName=<%= Encode.forUriComponent(fieldNoteEform.getFormName()) %>…">
```
Note: for URL query-string components, `Encode.forUriComponent` is more appropriate than `forHtmlAttribute`.

### Backwards compatibility
- Legitimate form names (alphanumeric, spaces, hyphens) render identically after encoding.
- URL-encoding form names in href contexts is strictly more correct (spaces become `+` etc.) — any code that parsed `formName` on the receiving end should already handle URL-decoding.
- Verify `fieldNoteReport.do` URL-decodes its `formName` parameter (it should, per standard servlet behavior).

### Acceptance criteria
- [ ] Add OWASP encoder import if not present
- [ ] Wrap body-context emissions with `Encode.forHtml()`
- [ ] Wrap URL-context emission (line 146) with `Encode.forUriComponent()`
- [ ] Manual smoke: admin eform list renders; clicking a row navigates to the report correctly

### Related alerts
Adversarial discovery; not scanner-flagged. Same XSS pattern as others in this bundle.

---

## Fix 6 — Encode audit-log fields in `demographicAudit.jsp`

**File**: `src/main/webapp/demographic/demographicAudit.jsp`
**Lines**: 114, 116, 118

### Problem
```jsp
<td><%=providerDao.getProviderName(log.getProviderNo())%></td>
<td><%=log.getAction()%></td>
<td><%=log.getContent()%></td>
```
Raw emission of audit-log action/content fields. Lines 120 and 122 of the **same file** already use `Encode.forHtml()` — this is an incomplete encoding sweep, not a designed choice. An attacker who triggers an audit-logged action with a crafted content string delivers stored XSS to any admin viewing the audit log.

### Fix
```jsp
<td><%= Encode.forHtml(providerDao.getProviderName(log.getProviderNo())) %></td>
<td><%= Encode.forHtml(log.getAction()) %></td>
<td><%= Encode.forHtml(log.getContent()) %></td>
```

### Backwards compatibility
- Consistent with neighboring lines 120/122 already encoded — brings the file to a uniform encoding pattern.
- Audit-log content is typically system-generated strings (action names, demographic IDs); encoding is a pure no-op for those.

### Acceptance criteria
- [ ] Wrap lines 114, 116, 118 with `Encode.forHtml()`
- [ ] Verify file is now internally consistent (no remaining raw `<%= … %>` of logged values)
- [ ] Manual smoke: open demographic audit log → renders correctly

### Related alerts
Adversarial discovery; not scanner-flagged. Same pattern.

---

## Fix 7 — Pre-encode HRM report content in `displayHRMReport.jsp`

**File**: `src/main/webapp/hospitalReportManager/displayHRMReport.jsp`
**Line**: 522

### Problem
```jsp
<%= hrmReport.getFirstReportTextContent().replaceAll("\n", "<br />") %>
```
HRM (Hospital Report Manager) content is ingested from external hospital systems — the most untrusted source in the EMR. Raw emission into HTML body with only newline-to-`<br>` substitution. A malicious or compromised upstream hospital report can deliver stored XSS.

### Fix
Encode before the newline substitution (matches the pattern already in use at `casemgmt/displayNote.jsp` and `lab/CA/ALL/labDisplayAjax.jsp`):
```jsp
<%= Encode.forHtml(hrmReport.getFirstReportTextContent()).replaceAll("\n", "<br />") %>
```

### Backwards compatibility
- Report text content is typically plain clinical prose — encoded output indistinguishable from current.
- HTML characters (`<`, `>`, `&`) in legitimate report content are now escaped (correct behavior; HRM content is not supposed to be HTML-styled).
- Newline-to-`<br>` substitution still produces visible line breaks.

### Acceptance criteria
- [ ] Add OWASP encoder import if not present
- [ ] Apply `Encode.forHtml()` before `.replaceAll`
- [ ] Manual smoke: open an HRM report → renders with line breaks preserved

### Related alerts
Adversarial discovery; not scanner-flagged despite being a high-trust-boundary XSS sink.

---

## Overall acceptance criteria

- [ ] All 7 fixes applied as a single PR or grouped small PRs (preference: one PR for easier review)
- [ ] No new dependencies; no new Maven entries
- [ ] No database migration
- [ ] Existing test suite (`make install --run-modern-tests`) passes
- [ ] Manual smoke test of affected flows: document viewer, prescription, note unlock, admin eforms, OAuth admin, demographic audit log, HRM viewer
- [ ] `make install` completes cleanly

## Test plan

- [ ] Run `make install --run-modern-tests` — existing tests pass
- [ ] Smoke: navigate a multi-page document (Fix 1)
- [ ] Smoke: prescribe a drug with a standard unit (Fix 2)
- [ ] Adversarial: create a clinical note with body `<b>bold</b>` → text displays literally, does not bold (Fix 3)
- [ ] Smoke: view OAuth client list as admin (Fix 4)
- [ ] Smoke: navigate the admin eform list (Fix 5)
- [ ] Smoke: open demographic audit log (Fix 6)
- [ ] Smoke: open an HRM report with multi-line content (Fix 7)

## Dismissing related code-scanning alerts

After merge:
- Closes `java/XSS` REVIEW alerts on `MultiPageDocDisplay.jsp` (4)
- Closes `java/RegexInjection` REVIEW alerts on `RxWriteScript2Action.java` (2)
- Closes ~4 adversarial-discovery XSS sinks not in the original scanner output

## Why this grouping

All 7 items follow the same shape as issue #1634:
- Single-file changes with clear before/after
- One pattern: add an `Encode.*` or `Pattern.quote` call
- No API, DB, dependency, or integration surface touched
- Zero runtime change for well-formed input
- Each fix independently verifiable
- Risk of regression: negligible (encoding idempotent on benign content)

## Out of scope (handle separately)

Related issues intentionally not bundled here because they require broader discussion or larger diff:
- Additional XSS JSPs discovered during adversarial audit but requiring multi-field fixes: `tickler/ticklerDemoMain.jsp`, `casemgmt/ticklers.jsp`, `oscarRx/Preview2.jsp`, `lab/CA/BC/index.jsp`, `WEB-INF/jsp/caisicore/DefaultEncounterIssueList.jsp`
- Reflected XSS via `<c:url>` + `escapeXml="false"` pattern in `casemgmt/search.jsp` and `casemgmt/CaseManagementView.jsp` — needs discussion of the JSTL anti-pattern
- Systematic grep-based sweep for `<%= \w+\.\w+\(\) %>` without an `Encode.` wrapper — a broader hardening project

---

🤖 Based on multi-pass adversarial validation of GitHub code-scanning alerts (April 2026). Companion to #1634.
